### PR TITLE
Magic square generator

### PR DESCRIPTION
**Pull request by juansebastianl**
Adding a (normal) magic square generator to the generators package. 

[Magic squares](https://en.wikipedia.org/wiki/Magic_square) are quadratic constrained integer problems of size $n^2$ with no objective.

### CODE REVIEW

File: dimod/generators/magic_square.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        variables[(i,j)] = Integer(f"var_{i}_{j}", lower_bound= 1)
+    
+    # the sum across every row, column, diagonal
+    constraint_sum = Integer(f"sum", lower_bound= 1)
+    
+    cqm = ConstrainedQuadraticModel()
+    
+    for i in range(size):
+        # Every row must equal the same value
+        cqm.add_constraint_from_comparison(quicksum([variables[(i,j)]**power for j in range(size)]) - constraint_sum == 0)
```
Comment (arcondello) [inline dimod/generators/magic_square.py:50]: Can also using dict's conversion from tuple indexing
```suggestion
        cqm.add_constraint_from_comparison(quicksum([variables[i, j]**power for j in range(size)]) - constraint_sum == 0)
```
here and elsewhere.

File: dimod/generators/magic_square.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from itertools import product
+
+from dimod.constrained import ConstrainedQuadraticModel
+from dimod.binary.binary_quadratic_model import quicksum
+from dimod.quadratic import Integer 
+
```
Comment (arcondello) [inline dimod/generators/magic_square.py:20]:
```suggestion

__all__ = ["magic_square"]

```

**Review by arcondello — COMMENTED** [review]
Needs a release note and to be added to the docs. Probably also want to import it into the [dimod.generators namespace](https://github.com/dwavesystems/dimod/blob/main/dimod/generators/__init__.py).

**Comment by arcondello** [comment]
Also needs tests.

File: dimod/generators/magic_square.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            If power is greater than 1 then the value of the variable represents the base, so the magic square is satisfied 
+            by var_i_j**power . 
+    """
+    
+    if power not in [1, 2]: 
+        raise ValueError(f"power must be either 1 or 2, recieved {power}")
+    
+    variables = {}
+    for i, j in product(range(size), range(size)): 
+        variables[(i,j)] = Integer(f"var_{i}_{j}", lower_bound= 1)
```
[the remark this answers is not in the document]
Comment (juansebastianl) [reply dimod/generators/magic_square.py:41]: Not sure what part of the spacing is off

[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/generators/magic_square.py:41]: So, for instance in this case it should be
```suggestion
        variables[(i, j)] = Integer(f"var_{i}_{j}", lower_bound=1)
```
Space between `i` and `j` and no space around `=` for the keyword argument. See [pep8](https://peps.python.org/pep-0008/#whitespace-in-expressions-and-statements). Recommend using a linter if you're not already.

**Review by arcondello — APPROVED** [review]
LGTM

**Pull request merged**